packages/filesystem/src/node/filesystem-watcher-dispatcher.ts

```diff
@@ -0,0 +1,82 @@
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import {
+    FileSystemWatcherClient, FileSystemWatcherClient2,
+    DidFilesChangedParams2, FileSystemWatcherErrorParams2
+} from '../common/filesystem-watcher-protocol';
+
+/**
+ * This component routes watch events to the right clients.
+ */
+@injectable()
+export class FileSystemWatcherServer2Dispatcher implements FileSystemWatcherClient2 {
+
+    /**
+     * Mapping of `clientId` to actual clients.
+     */
+    protected readonly clients = new Map<number, FileSystemWatcherClient>();
+
+    onDidFilesChanged2(event: DidFilesChangedParams2): void {
+        for (const client of this.iterRegisteredClients(event.clients)) {
+            client.onDidFilesChanged(event);
+        }
+    }
+
+    onError2(event: FileSystemWatcherErrorParams2): void {
+        for (const client of this.iterRegisteredClients(event.clients)) {
+            client.onError();
+        }
+    }
+
+    /**
+     * Listen for events targeted at `clientId`.
+     */
+    registerClient(clientId: number, client: FileSystemWatcherClient): void {
+        if (this.clients.has(clientId)) {
+            console.warn(`FileSystemWatcherServer2Dispatcher: a client was already registered! clientId=${clientId}`);
+        }
+        this.clients.set(clientId, client);
+    }
+
+    unregisterClient(clientId: number): void {
+        if (!this.clients.has(clientId)) {
+            console.warn(`FileSystemWatcherServer2Dispatcher: tried to remove unknown client! clientId=${clientId}`);
+        }
+        this.clients.delete(clientId);
+    }
+
+    /**
+     * Only yield registered clients for the given `clientIds`.
+     *
+     * If clientIds is empty, will return all clients.
+     */
+    protected *iterRegisteredClients(clientIds?: number[]): Iterable<FileSystemWatcherClient> {
+        if (!Array.isArray(clientIds) || clientIds.length === 0) {
+            // If we receive an event targeted to "no client",
+            // interpret that as notifying all clients:
+            yield* this.clients.values();
+        } else {
+            for (const clientId of clientIds) {
+                const client = this.clients.get(clientId);
+                if (typeof client !== 'undefined') {
```

Comment:
Why do not you use `if (client)`. I think you're confused with the [`ReferenceError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Not_defined).
It's a good practice to do something like `typeof process !== 'undefined'` to make sure you do not have a reference error at runtime (because your code runs in browser), but you do not need to do it every single time, such as for this code. `const client` defines the variable.

Reply:
Right, I can directly test for `client !== undefined`, just an habit of using `typeof`. Will change.